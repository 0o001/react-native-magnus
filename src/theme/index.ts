export { ThemeContext } from './themeContext';
export { withTheme } from './withTheme';
export { ThemeProvider, ThemeProviderProps } from './themeProvider.component';

export {
  ThemeType,
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from './type';
