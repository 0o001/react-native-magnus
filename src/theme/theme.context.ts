import * as React from 'react';

import { ThemeType } from './type';

const defaultValue: ThemeType = {};

export interface ThemeContextType {
  theme: ThemeType;
  setTheme?: (theme?: ThemeType) => void;
}

export const THEME_CONTEXT_DEFAULT_VALUE = {
  theme: defaultValue,
};

export const ThemeContext: React.Context<ThemeContextType> = React.createContext(
  THEME_CONTEXT_DEFAULT_VALUE
);
