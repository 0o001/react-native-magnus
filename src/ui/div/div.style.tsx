import { StyleSheet } from "react-native";

import {
  getThemeProperty,
  createFlexStyles,
  createPositionStyle,
  createSpacingStyles,
  createBorderWidthStyles,
  createBorderColorStyles,
  createBorderRadiusStyles
} from "../../theme/theme.service";

/**
 * computed style
 *
 * @param theme
 * @param props
 */
export const getStyle = (theme: any, props: any) => {
  const computedStyle: any = {};

  computedStyle.div = {
    flexDirection: props.flexDir,
    flexWrap: props.flexWrap,
    alignItems: props.alignItems,
    justifyContent: props.justifyContent,
    alignSelf: props.alignSelf,
    backgroundColor: getThemeProperty(theme.colors, props.bg, "transparent"),
    ...createFlexStyles(props),
    ...createPositionStyle(props),
    ...createBorderWidthStyles(props),
    ...createSpacingStyles(props, theme.spacing),
    ...createBorderColorStyles(props, theme.colors),
    ...createBorderRadiusStyles(props, theme.borderRadius)
  };

  computedStyle.image = {
    ...createBorderRadiusStyles(props, theme.borderRadius)
  };

  if (props.shadow) {
    computedStyle.div = {
      ...computedStyle.div,
      ...theme.shadow[props.shadow],
      shadowColor: getThemeProperty(theme.colors, props.shadowColor, "white")
    };
  }

  if (props.h) {
    computedStyle.div = {
      ...computedStyle.div,
      height: props.h
    };
  }

  if (props.w) {
    computedStyle.div = {
      ...computedStyle.div,
      width: props.w
    };
  }

  if (props.minH) {
    computedStyle.div = {
      ...computedStyle.div,
      minHeight: props.minH
    };
  }

  if (props.minW) {
    computedStyle.div = {
      ...computedStyle.div,
      minWidth: props.minW
    };
  }

  if (props.opacity) {
    computedStyle.div = {
      ...computedStyle.div,
      opacity: props.opacity
    };
  }

  // merging style props to computed style
  if (props.style) {
    computedStyle.div = {
      ...computedStyle.div,
      ...props.style
    };
  }

  return StyleSheet.create(computedStyle);
};
