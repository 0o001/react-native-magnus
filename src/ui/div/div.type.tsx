import {
  ViewProps as RNViewProps,
  ImageSourcePropType as RNImageSourcePropType,
} from 'react-native';

import {
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
  ShadowPropsType,
  ResponsiveType,
  BaseBgModeType,
  BaseJustifyType,
  BaseAlignSelfType,
  BaseAlignItemsType,
  BaseFlexWrapType,
  BaseFlexDirectionType,
  BasePositionType,
  BaseOverflowType,
  BaseBorderStyleType,
} from '../../theme';

export interface DivProps
  extends RNViewProps,
    BorderPropsType,
    SpacingPropsType,
    RoundedPropsType,
    ShadowPropsType {
  h?: ResponsiveType<string | number>;
  w?: ResponsiveType<string | number>;
  bg?: ResponsiveType<string>;
  minH?: ResponsiveType<string | number>;
  minW?: ResponsiveType<string | number>;
  maxH?: ResponsiveType<string | number>;
  maxW?: ResponsiveType<string | number>;
  bgImg?: RNImageSourcePropType;
  bgMode?: ResponsiveType<BaseBgModeType>;
  flex?: ResponsiveType<number>;
  justifyContent?: ResponsiveType<BaseJustifyType>;
  alignSelf?: ResponsiveType<BaseAlignSelfType>;
  alignItems?: ResponsiveType<BaseAlignItemsType>;
  flexDir?: ResponsiveType<BaseFlexDirectionType>;
  flexWrap?: ResponsiveType<BaseFlexWrapType>;
  position?: ResponsiveType<BasePositionType>;
  overflow?: ResponsiveType<BaseOverflowType>;
  borderStyle?: ResponsiveType<BaseBorderStyleType>;
  top?: ResponsiveType<number>;
  right?: ResponsiveType<number>;
  bottom?: ResponsiveType<number>;
  left?: ResponsiveType<number>;
  opacity?: ResponsiveType<number>;
  children?: React.ReactNode[] | React.ReactNode;
  row?: ResponsiveType<boolean>;
  zIndex?: ResponsiveType<number>;
}
