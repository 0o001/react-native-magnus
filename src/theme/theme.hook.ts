import React, { useState } from "react";

import { ThemeType } from "./type";
import { computeTheme } from "./theme.service";
import { ThemeContextType } from "./theme.context";

export const useTheme = (initialTheme?: ThemeType): ThemeContextType => {
  const [theme, setTheme] = useState(computeTheme(initialTheme));

  const setCurrentTheme = React.useCallback((newTheme?: ThemeType): void => {
    setTheme(computeTheme(newTheme));
  }, []);

  return {
    theme,
    setTheme: setCurrentTheme
  };
};
