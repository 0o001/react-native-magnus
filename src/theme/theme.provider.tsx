import * as React from "react";

import { ThemeType } from "./type";

import { ThemeContext } from "./theme.context";
import { useTheme } from "./theme.hook";

export interface ThemeProviderProps {
  theme?: ThemeType;
  children?: React.ReactNode;
}

export const ThemeProvider: React.FunctionComponent<ThemeProviderProps> = (
  props: ThemeProviderProps
) => {
  const { theme: themeProp, children } = props;
  const theme = useTheme(themeProp);

  return (
    <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>
  );
};
