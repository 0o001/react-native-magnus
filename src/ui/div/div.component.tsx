import * as React from "react";
import { useContext } from "react";
import {
  View as RNView,
  ImageBackground as RNImageBackground
} from "react-native";

import { DivProps } from "./div.type";
import { getStyle } from "./div.style";
import { ThemeContext } from "../../theme";

const Div: React.FunctionComponent<DivProps> = (props: DivProps) => {
  const {
    bg,
    h,
    w,
    m,
    mt,
    mr,
    mb,
    ml,
    p,
    pr,
    pt,
    pb,
    pl,
    minH,
    minW,
    maxW,
    maxH,
    position,
    style,
    flexDir,
    row,
    rounded,
    roundedTop,
    roundedRight,
    roundedBottom,
    roundedLeft,
    children,
    bgImg,
    bgMode,
    alignItems,
    justifyContent,
    borderColor,
    borderBottomColor,
    borderLeftColor,
    borderTopColor,
    borderRightColor,
    borderWidth,
    borderLeftWidth,
    borderRightWidth,
    borderBottomWidth,
    borderTopWidth,
    flexWrap,
    shadow,
    shadowColor,
    opacity,
    overflow,
    top,
    left,
    right,
    bottom,
    zIndex,
    ...rest
  } = props;
  const theme = useContext(ThemeContext);
  const computedStyle = getStyle(theme, props);

  if (bgImg) {
    return (
      <RNImageBackground
        source={bgImg}
        style={computedStyle.div}
        resizeMode={props.bgMode}
        imageStyle={computedStyle.image}
        {...rest}
      >
        {children}
      </RNImageBackground>
    );
  }

  return (
    <RNView style={computedStyle.div} {...rest}>
      {children}
    </RNView>
  );
};

Div.defaultProps = {
  bg: "transparent",
  flexDir: "column",
  flexWrap: "nowrap",
  rounded: "none",
  shadow: 0,
  shadowColor: "gray900",
  position: "relative",
  bgMode: "cover",
  pointerEvents: "auto",
  row: false,
  borderStyle: "solid"
};

export { Div };
