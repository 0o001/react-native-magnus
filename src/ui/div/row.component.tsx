import * as React from "react";
import { useContext } from "react";
import {
  View as RNView,
  ImageBackground as RNImageBackground
} from "react-native";

import { DivProps } from "./div.type";
import { getStyle } from "./div.style";
import { ThemeContext } from "../../theme";

const Row: React.FunctionComponent<DivProps> = (props: DivProps) => {
  const {
    h,
    w,
    m,
    mt,
    mr,
    mb,
    ml,
    p,
    pr,
    pt,
    pb,
    pl,
    minH,
    minW,
    position,
    style,
    flexDir,
    rounded,
    roundedTop,
    roundedRight,
    roundedBottom,
    roundedLeft,
    children,
    bgImg,
    bgMode,
    alignItems,
    justifyContent,
    borderColor,
    borderBottomColor,
    borderLeftColor,
    borderTopColor,
    borderRightColor,
    borderWidth,
    borderLeftWidth,
    borderRightWidth,
    borderBottomWidth,
    borderTopWidth,
    flexWrap,
    shadow,
    shadowColor,
    ...rest
  } = props;
  const theme = useContext(ThemeContext);
  const computedStyle = getStyle(theme, props);

  if (bgImg) {
    return (
      <RNImageBackground
        source={bgImg}
        style={computedStyle.div}
        resizeMode={props.bgMode}
        imageStyle={computedStyle.image}
        {...rest}
      >
        {children}
      </RNImageBackground>
    );
  }

  return (
    <RNView style={computedStyle.div} {...rest}>
      {children}
    </RNView>
  );
};

Row.defaultProps = {
  bg: "white",
  flexDir: "row",
  flexWrap: "nowrap",
  rounded: "none",
  alignItems: "center",
  shadow: 0,
  shadowColor: "gray900",
  position: "relative",
  bgMode: "cover"
};

export { Row };
