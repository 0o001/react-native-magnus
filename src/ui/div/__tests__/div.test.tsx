import React from 'react';
import renderer from 'react-test-renderer';

import { Div } from '../div.component';
import { ThemeProvider } from '../../../theme';

describe('Div Component', () => {
  it('should render without issues', () => {
    const instance = renderer.create(
      <ThemeProvider>
        <Div />
      </ThemeProvider>,
    );

    expect(instance.toJSON()).toMatchSnapshot();

    const divInstance = instance.root.findByType(Div);
    expect(divInstance.props.children).toBe(undefined);
  });

  it('should render with props', () => {
    const instance = renderer.create(
      <ThemeProvider>
        <Div bg="red500" />
      </ThemeProvider>,
    );

    const divInstance = instance.root.findByType(Div);
    expect(divInstance.props.bg).toBe('red500');
  });
});
