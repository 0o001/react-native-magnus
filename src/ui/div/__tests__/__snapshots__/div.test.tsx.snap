// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Div Component should render without issues 1`] = `
<View
  style={
    Object {
      "alignItems": "flex-start",
      "backgroundColor": "#FFFFFF",
      "borderRadius": 0,
      "flexDirection": "column",
      "flexWrap": "nowrap",
      "justifyContent": undefined,
      "position": "relative",
    }
  }
/>
`;
